File: day5/d5.py
from input import realData
from input import testData
from input import print_in_rows

def get_data ():
    input_data = realData
    input = input_data.strip().split("\n")
    inputLen = len(input)

    data = []

    for i in range(inputLen):
        rangeNums = input[i].split('-')
        start = int(rangeNums[0])
        end = int(rangeNums[1])
        data.append([start, end])
    
    sorted_data = sorted(data)
    return sorted_data

# we have a sorted list of ranges
# return all the numbers between the ranges (NO OVERLAP)
# how many numbers between each range, total, return

def calculate_range_values(data):
    total = 0
    for r_list in data:
        total += (r_list[1] -r_list[0]+1)
    return total

def merge_data(data):
    merged = []
    for start, end in sorted(data):  # already sorted, but we can sort again yall
        if not merged or start > merged[-1][1] + 1:  # no overlap/adjacent -> append that hoe
            merged.append([start, end])
        else:  # overlaps or touches
            merged[-1][1] = max(merged[-1][1], end)
    return merged


def main () : 
    data = get_data() # ingest data
    merged_data = merge_data(data) # merge data to accomodate gaps
    print(calculate_range_values(merged_data)) # calculate fresh produce

main()

File: day5/d5.js
import { readFileSync } from 'fs'
import { fileURLToPath } from 'url'
import path from 'path'

const ingestData = () => {
  const __filename = fileURLToPath(import.meta.url)
  const __dirname = path.dirname(__filename)
  return readFileSync(path.join(__dirname, 'josh.txt'), 'utf-8')
}

const parseRangeString = (str) => {
  const [start, end] = str.split('-').map((part) => Number(part.trim()))
  return { start, end }
}

const parseIdString = (str) => {
  return Number(str.trim())
}

const sortData = (input) => {
  // split by lines
  const lines = input.split(/\r?\n/)
  // organize the raw text data into two usable arrays
  let ranges = []
  let ids = []
  let checkRanges = true

  lines.forEach((line) => {
    // if line is just whitespace (indicating we're moving on to ids)
    if (checkRanges) {
      if (line.trim().length <= 0) {
        checkRanges = false
      } else {
        // add the range to the list
        ranges.push(parseRangeString(line))
      }
    } else {
      // add the id to the list
      if (line.trim().length > 0) ids.push(parseIdString(line))
    }
  })

  return { ranges, ids }
}

const checkIds = (ranges, ids) => {
  let numFresh = 0

  for (let i = 0; i < ids.length; i++) {
    // for each id loop ranges until you fit in one then break
    let j = 0
    while (j < ranges.length) {
      if (ids[i] >= ranges[j].start && ids[i] <= ranges[j].end) {
        numFresh++
        break
      }
      j++
    }
  }

  return numFresh
}

// const main = () => {
//   // 1. ingest data -> get ranges and ids to check
//   const input = ingestData()
//   // 2. organize input
//   const { ranges, ids } = sortData(input)
//   // 3. check each id against ranges
//   const numFresh = checkIds(ranges, ids)
//   // 4. return num fresh
//   console.log(`Number of Fresh Ingredients: ${numFresh}`)
// }

const mergeRanges = (ranges) => {
  // sort by start so we can sweep once and merge overlaps/adjacent ranges
  const sorted = [...ranges].sort((a, b) => a.start - b.start)
  const merged = []

  for (const range of sorted) {
    const last = merged[merged.length - 1]
    if (!last || range.start > last.end + 1) {
      merged.push({ ...range })
    } else {
      last.end = Math.max(last.end, range.end)
    }
  }

  return merged
}

const calculateNumFreshFromRanges = (ranges) => {
  // merge overlaps so each id is counted once
  const merged = mergeRanges(ranges)

  return merged.reduce((total, { start, end }) => {
    return total + (end - start + 1) // +1 because ranges are inclusive
  }, 0)
}

// pt 2
const partTwo = () => {
  // 1. ingest data
  const input = ingestData()
  // 2. organize input
  const { ranges, ids } = sortData(input)
  // 3. check ranges for num ids (beware of overlap)
  const numFresh = calculateNumFreshFromRanges(ranges)
  // 4. return num fresh
  console.log(numFresh)
}

partTwo()
